File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def solution(arr):
    def compress(x, y, size):
        first = arr[x][y]
        for i in range(x, x + size):
            for j in range(y, y + size):
                if arr[i][j] != first:
                    half = size // 2
                    res1 = compress(x, y, half)
                    res2 = compress(x, y + half, half)
                    res3 = compress(x + half, y, half)
                    res4 = compress(x + half, y + half, half)
                    return [sum(x) for x in zip(res1, res2, res3, res4)]
        return [1, 0] if first == 0 else [0, 1]
    return compress(0, 0, len(arr))

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.js
function solution(arr) {
    function compress(x, y, size) {
        const first = arr[x][y];
        let allSame = true;

        for (let i = x ; i < x + size ; i++) {
            for (let j = y ; j < y + size ; j++) {
                if (arr[i][j] !== first) {
                    allSame = false;
                    break;
                }
            }
            if (!allSame) break;
        }

        if (allSame) return first === 0 ? [1, 0] : [0, 1];

        const half = size / 2;
        const res1 = compress(x, y, half);
        const res2 = compress(x, y + half, half);
        const res3 = compress(x + half, y, half);
        const res4 = compress(x + half, y + half, half);
        return res1.map((val, idx) => val + res2[idx] + res3[idx] + res4[idx]);
  }

  return compress(0, 0, arr.length);
}
